src/engine/PolarityBasedDivider.cpp

```diff
@@ -0,0 +1,112 @@
+
+**/
+
+#include "Debug.h"
+#include "EngineState.h"
+#include "FloatUtils.h"
+#include "PolarityBasedDivider.h"
+#include "MStringf.h"
+#include "PiecewiseLinearCaseSplit.h"
+
+PolarityBasedDivider::PolarityBasedDivider( std::shared_ptr<IEngine> engine )
+    : _engine( std::move( engine ) )
+{
+};
+
+void PolarityBasedDivider::createSubQueries( unsigned numNewSubqueries, const
+                                             String queryIdPrefix, const
+                                             PiecewiseLinearCaseSplit
+                                             &previousSplit, const unsigned
+                                             timeoutInSeconds, SubQueries
+                                             &subQueries )
+{
+    unsigned numBisects = (unsigned)log2( numNewSubqueries );
+
+    List<PiecewiseLinearCaseSplit *> splits;
+    auto split = new PiecewiseLinearCaseSplit();
+    *split = previousSplit;
+    splits.append( split );
+
+    for ( unsigned i = 0; i < numBisects; ++i )
+    {
+        List<PiecewiseLinearCaseSplit *> newSplits;
+        for ( const auto &split : splits )
+        {
+            PiecewiseLinearConstraint *pLConstraintToSplit =
+                getPLConstraintToSplit( *split );
+            if ( pLConstraintToSplit == NULL )
+            {
+                auto newSplit = new PiecewiseLinearCaseSplit();
+                *newSplit = *split;
+                newSplits.append( newSplit );
+            }
+            else
+            {
+                auto caseSplits = pLConstraintToSplit->getCaseSplits();
+                for ( const auto &caseSplit : caseSplits )
+                {
+                    auto newSplit = new PiecewiseLinearCaseSplit();
+                    *newSplit = *split;
+                    for ( const auto &tightening : caseSplit.getBoundTightenings() )
+                        newSplit->storeBoundTightening( tightening );
+                    newSplits.append( newSplit );
+                }
+            }
+            delete split;
+        }
+        splits = newSplits;
+    }
+
+    unsigned queryIdSuffix = 1; // For query id
+    // Create a new subquery for each newly created input region
+    for ( const auto &split : splits )
+    {
+        // Create a new query id
+        String queryId;
+        if ( queryIdPrefix == "" )
+            queryId = queryIdPrefix + Stringf( "%u", queryIdSuffix++ );
+        else
+            queryId = queryIdPrefix + Stringf( "-%u", queryIdSuffix++ );
+
+        // Construct the new subquery and add it to subqueries
+        SubQuery *subQuery = new SubQuery;
+        subQuery->_queryId = queryId;
+        subQuery->_split.reset(split);
+        subQuery->_timeoutInSeconds = timeoutInSeconds;
+        subQueries.append( subQuery );
+    }
+}
+
+PiecewiseLinearConstraint *PolarityBasedDivider::getPLConstraintToSplit
+( const PiecewiseLinearCaseSplit &split )
+{
+    EngineState *engineStateBeforeSplit = new EngineState();
```

Comment:
I had a hard time following the flow here. First you apply one split; and then you pick the best split after the first split?

Reply:
If we want to create 4 subqueries, we do the following:
First:
  r1_inactive /\ r1_inactive = getPLConstraintToSplit( emptySplit )
Then:
 r2_inactive /\ r2_active = getPLConstraintToSplit( r1_inactive )
 r3_inactive /\ r3_active = getPLConstraintToSplit( r1_active )

In the end, we get
r1_inactive /\  r2_inactive
r1_inactive /\ r2_active
r1_active /\ r3_inactive
r1_active /\ r3_active